make_abi_library.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python
+
+import sys
+import fileinput
+from cparser import *
+
+# cache for "is this type a function pointer?"
+global POINTER_CACHE
+POINTER_CACHE = {}
+# cache for "is this type a function?"
+global FUNCTION_CACHE
+FUNCTION_CACHE = {}
+# statistics useful for sanity checking
+global STATS
+STATS = {}
+
+
+def has_function(args):
```

Comment:
I think what you should add a `seen` set argument to use alongside the global `FUNCTION_CACHE`. The reason why is that the initial assumption of `FUNCTION_CACHE[arg] = False` is there to stop recursion, but you don't want to prematurely determine that something else doesn't have a function pointer. For example:

```c
struct X;
typedef struct X Y;
struct X {
  Y *next;
  void (*foo)(void);
};
void bar(struct X *z);
void baz(Y *z);
````

I think in the above case, `baz` will be treated as not taking a function pointer, when really it does. So what you want is something like:

```python
def has_function(*args, seen=None):
  global FUNCTION_CACHE
  if seen is None:
    seen = set()
  for arg in args:
    # check FUNCTION_CACHE here, return whatever is in the FUNCTION_CACHE if present
    # check seen cache here, return False if seen
    # add to seen cache, but NOT function cache
    # if statements as below, passing `seen` down into calls of `has_function`, updating FUNCTION_CACHE with the result to be returned
```

By using `*args` instead of `args`, you can do `has_function(arg.ctype, seen=seen)`, and in the tuple case, do `has_function(*arg, seen=seen)`